### DISCUSSION TITLE
RavenDB.Embedded in library / not main project

### DISCUSSION THREAD

**Discussion by cjbor** [ravendb/ravendb#15725] [Q&A] [answered]
I am using the RavenDB.Embedded package in my project.
To seperate concerns, I put the RavenDB logic into its own persistence project.

I now have persistenceLib -> sessionLib -> UiCommonLib -> UiPlatformProj

First Problem: When using nuget for part of this the RavenDBServer doesnt get extracted anymore.
persistenceLib + sessionLib on nuget server 
UiCommonLib + UiPlatformProj as local project reference
The Raven.Embedded.dll and Raven.Client.dll get copied without problems however

Second Problem: Calling EmbeddedServer.Instance.GetDocumentStore(databaseName) results in an exception.
exception: Raven.Client.Exceptions.ConcurrencyException: Database 'Local' already exists!
(GetDocumentStore should only create a new database if it doesnt exists as far as I know)

Third Problem: Changes API doesnt trigger as Expected.
I will do some more investigation but dont have any details yet.

**Comment by ayende** [comment]
You need to have a reference from the executable, since RavenDB needs tom binary files that won't always be copied if you are using transitive references.
Note that you can just have the reference there, not actually call it from there. 

For the other issue, please post the full error and the calling code.

**Comment by cjbor** [comment]
The Exception:
```
Raven.Client.Exceptions.ConcurrencyException: Database 'Local' already exists!
   at Raven.Client.Exceptions.ExceptionDispatcher.ThrowConflict(ExceptionSchema schema, BlittableJsonReaderObject json) in C:\Builds\RavenDB-Stable-5.4\54027\src\Raven.Client\Exceptions\ExceptionDispatcher.cs:line 197
   at Raven.Client.Exceptions.ExceptionDispatcher.Throw(JsonOperationContext context, HttpResponseMessage response, Action`1 additionalErrorInfo) in C:\Builds\RavenDB-Stable-5.4\54027\src\Raven.Client\Exceptions\ExceptionDispatcher.cs:line 89
```
Yes the database 'Local' already exists. But why is this an exception? I sadly cant get that much information from the callstack.

EmbeddedRavenConnection.cs
```
public class EmbeddedRavenConnection : RavenConnection
{
    private readonly IDocumentStore _documentStore;
    
    public EmbeddedRavenConnection(string databaseName)
    {
        EmbeddedServer.Instance.StartServer();
        _documentStore = EmbeddedServer.Instance.GetDocumentStore(databaseName);
        _documentStore.Initialize();
    }

    internal override IDocumentStore GetDocumentStore()
    {
        return _documentStore;
    }
}
```

RavenStorage.cs
```
public class RavenStorage : Storage
{
    private readonly IDocumentStore _documentStore;
    private readonly IDocumentSession _session;
    
    public RavenStorage(RavenConnection ravenConnection)
    {
        _documentStore = ravenConnection.GetDocumentStore();
        _session = _documentStore.OpenSession();
    }
    
    public override void Store(object entity)
    {
        _session.Store(entity);
        _session.SaveChanges();
    }

    public override void Save()
    {
        _session.SaveChanges();
    }

    public override Subscription<T> CreateSubscription<T>(T obj)
    {
        return new RavenSubscription<T>(obj, _documentStore, _session);
    }
}
```

**Reply by ayende** [reply]
Are you stopping on the exception that is thrown?
This is handled here:

https://github.com/ravendb/ravendb/blob/1dda88948eb78d8be12cc7bec6d6e5be744c3493/src/Raven.Embedded/EmbeddedServer.cs#L174

If you hit F5, it should continue normally

**Reply by cjbor** [reply]
Yes im stopped at an Exception.

https://github.com/ravendb/ravendb/blob/1dda88948eb78d8be12cc7bec6d6e5be744c3493/src/Raven.Client/Exceptions/ExceptionDispatcher.cs#L197

There to be exact.

For some reason I also wasnt able to catch the exception.

```
try
{
    _documentStore = EmbeddedServer.Instance.GetDocumentStore(databaseName);
}
catch (Exception e)
{

}
```

This still resulted in the application halting at the mentioned exception.

**Reply by ayende** [reply]
You are in the debugger, stopped on a *handled exception*.
Hit F5, it will work

**Reply by cjbor** [reply]
After pressing F5 (continue) I get thrown to ExceptionDispatchInfo.cs in System.Runtime.ExceptionServices

![grafik](https://user-images.githubusercontent.com/21296630/215491509-7d8b3cfc-17ae-441b-852d-f1565c045ea3.png)

after clicking F5 (continue) about 15 more times the software eventually continues.
I had the EmbeddedServer working without problems before. Something in my setup must have changed.
I cant really say why it happens but I cant work with the EmbeddedServer like that.
For now I switched to the standalone server as that one works flawlessly.

**Reply by ayende** [reply]
You can mute the exception

**Comment by cjbor** [comment, marked as answer]
I managed to resolve the third problem by using:

```
var subscription = _documentStore.Changes();
await subscription.EnsureConnectedNow();
var observableWithTask = subscription.ForDocument(id);
observableWithTask.Subscribe(change =>
{
    if (change.Type == DocumentChangeTypes.Put)
    {
        var item = _session.Load<T>(id);
        ItemChanged?.Invoke(this, item);
    }
});
await observableWithTask.EnsureSubscribedNow();
```

I wasnt calling EnsureSubscribedNow before and the change to the database literally happend in the next line.

**Reply by ayende** [reply]
Yes, `Changes` are inherently racy, this is the way to handle them

**Comment by cjbor** [comment]
Thanks for the support
